File: compiler/packages/babel-plugin-react-compiler/src/Validation/ValidateHooksUsage.ts
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import * as t from '@babel/types';
import {
  CompilerError,
  CompilerErrorDetail,
  ErrorSeverity,
} from '../CompilerError';
import {computeUnconditionalBlocks} from '../HIR/ComputeUnconditionalBlocks';
import {isHookName} from '../HIR/Environment';
import {
  HIRFunction,
  IdentifierId,
  Place,
  SourceLocation,
  getHookKind,
} from '../HIR/HIR';
import {
  eachInstructionLValue,
  eachInstructionOperand,
  eachTerminalOperand,
} from '../HIR/visitors';
import {assertExhaustive} from '../Utils/utils';

/**
 * Represents the possible kinds of value which may be stored at a given Place during
 * abstract interpretation. The kinds form a lattice, with earlier items taking
 * precedence over later items (see joinKinds()).
 */
enum Kind {
  // A potential/known hook which was already used in an invalid way
  Error = 'Error',

  /*
   * A known hook. Sources include:
   * - LoadGlobal instructions whose type was inferred as a hook
   * - PropertyLoad, ComputedLoad, and Destructuring instructions
   *   where the object is a KnownHook
   * - PropertyLoad, ComputedLoad, and Destructuring instructions
   *   where the object is a Global and the property name is hook-like
   */
  KnownHook = 'KnownHook',

  /*
   * A potential hook. Sources include:
   * - LValues (other than LoadGlobal) where the name is hook-like
   * - PropertyLoad, ComputedLoad, and Destructuring instructions
   *   where the object is a potential hook or the property name
   *   is hook-like
   */
  PotentialHook = 'PotentialHook',

  // LoadGlobal values whose type was not inferred as a hook
  Global = 'Global',

  // All other values, ie local variables
  Local = 'Local',
}

function joinKinds(a: Kind, b: Kind): Kind {
  if (a === Kind.Error || b === Kind.Error) {
    return Kind.Error;
  } else if (a === Kind.KnownHook || b === Kind.KnownHook) {
    return Kind.KnownHook;
  } else if (a === Kind.PotentialHook || b === Kind.PotentialHook) {
    return Kind.PotentialHook;
  } else if (a === Kind.Global || b === Kind.Global) {
    return Kind.Global;
  } else {
    return Kind.Local;
  }
}

/*
 * Validates that the function honors the [Rules of Hooks](https://react.dev/warnings/invalid-hook-call-warning)
 * rule that hooks may only be called and not otherwise referenced as first-class values.
 *
 * Specifically this pass implements the following rules:
 * - Known hooks may only be called unconditionally, and cannot be used as first-class values.
 *   See the note for Kind.KnownHook for sources of known hooks
 * - Potential hooks may be referenced as first-class values, with the exception that they
 *   may not appear as the callee of a conditional call.
 *   See the note for Kind.PotentialHook for sources of potential hooks
 */
export function validateHooksUsage(fn: HIRFunction): void {
  const unconditionalBlocks = computeUnconditionalBlocks(fn);

  const errors = new CompilerError();
  const errorsByPlace = new Map<t.SourceLocation, CompilerErrorDetail>();

  function recordError(
    loc: SourceLocation,
    errorDetail: CompilerErrorDetail,
  ): void {
    if (typeof loc === 'symbol') {
      errors.pushErrorDetail(errorDetail);
    } else {
      errorsByPlace.set(loc, errorDetail);
    }
  }

  function recordConditionalHookError(place: Place): void {
    // Once a particular hook has a conditional call error, don't report any further issues for this hook
    setKind(place, Kind.Error);

    const reason =
      'Hooks must always be called in a consistent order, and may not be called conditionally. See the Rules of Hooks (https://react.dev/warnings/invalid-hook-call-warning)';
    const previousError =
      typeof place.loc !== 'symbol' ? errorsByPlace.get(place.loc) : undefined;

    /*
     * In some circumstances such as optional calls, we may first encounter a "hook may not be referenced as normal values" error.
     * If that same place is also used as a conditional call, upgrade the error to a conditonal hook error
     */
    if (previousError === undefined || previousError.reason !== reason) {
      recordError(
        place.loc,
        new CompilerErrorDetail({
          description: null,
          reason,
          loc: place.loc,
          severity: ErrorSeverity.InvalidReact,
          suggestions: null,
        }),
      );
    }
  }
  function recordInvalidHookUsageError(place: Place): void {
    const previousError =
      typeof place.loc !== 'symbol' ? errorsByPlace.get(place.loc) : undefined;
    if (previousError === undefined) {
      recordError(
        place.loc,
        new CompilerErrorDetail({
          description: null,
          reason:
            'Hooks may not be referenced as normal values, they must be called. See https://react.dev/reference/rules/react-calls-components-and-hooks#never-pass-around-hooks-as-regular-values',
          loc: place.loc,
          severity: ErrorSeverity.InvalidReact,
          suggestions: null,
        }),
      );
    }
  }
  function recordDynamicHookUsageError(place: Place): void {
    const previousError =
      typeof place.loc !== 'symbol' ? errorsByPlace.get(place.loc) : undefined;
    if (previousError === undefined) {
      recordError(
        place.loc,
        new CompilerErrorDetail({
          description: null,
          reason:
            'Hooks must be the same function on every render, but this value may change over time to a different function. See https://react.dev/reference/rules/react-calls-components-and-hooks#dont-dynamically-use-hooks',
          loc: place.loc,
          severity: ErrorSeverity.InvalidReact,
          suggestions: null,
        }),
      );
    }
  }

  const valueKinds = new Map<IdentifierId, Kind>();
  function getKindForPlace(place: Place): Kind {
    const knownKind = valueKinds.get(place.identifier.id);
    if (
      place.identifier.name !== null &&
      isHookName(place.identifier.name.value)
    ) {
      return joinKinds(knownKind ?? Kind.Local, Kind.PotentialHook);
    } else {
      return knownKind ?? Kind.Local;
    }
  }

  function visitPlace(place: Place): void {
    const kind = valueKinds.get(place.identifier.id);
    if (kind === Kind.KnownHook) {
      recordInvalidHookUsageError(place);
    }
  }

  function setKind(place: Place, kind: Kind): void {
    valueKinds.set(place.identifier.id, kind);
  }

  for (const param of fn.params) {
    const place = param.kind === 'Identifier' ? param : param.place;
    const kind = getKindForPlace(place);
    setKind(place, kind);
  }

  for (const [, block] of fn.body.blocks) {
    for (const phi of block.phis) {
      let kind: Kind =
        phi.id.name !== null && isHookName(phi.id.name.value)
          ? Kind.PotentialHook
          : Kind.Local;
      for (const [, operand] of phi.operands) {
        const operandKind = valueKinds.get(operand.id);
        /*
         * NOTE: we currently skip operands whose value is unknown
         * (which can only occur for functions with loops), we may
         * cause us to miss invalid code in some cases. We should
         * expand this to a fixpoint iteration in a follow-up.
         */
        if (operandKind !== undefined) {
          kind = joinKinds(kind, operandKind);
        }
      }
      valueKinds.set(phi.id.id, kind);
    }
    for (const instr of block.instructions) {
      switch (instr.value.kind) {
        case 'LoadGlobal': {
          /*
           * Globals are the one source of known hooks: they are either
           * directly a hook, or infer a Global kind from which knownhooks
           * can be derived later via property access (PropertyLoad etc)
           */
          if (getHookKind(fn.env, instr.lvalue.identifier) != null) {
            setKind(instr.lvalue, Kind.KnownHook);
          } else {
            setKind(instr.lvalue, Kind.Global);
          }
          break;
        }
        case 'LoadContext':
        case 'LoadLocal': {
          visitPlace(instr.value.place);
          const kind = getKindForPlace(instr.value.place);
          setKind(instr.lvalue, kind);
          break;
        }
        case 'StoreLocal':
        case 'StoreContext': {
          visitPlace(instr.value.value);
          const kind = joinKinds(
            getKindForPlace(instr.value.value),
            getKindForPlace(instr.value.lvalue.place),
          );
          setKind(instr.value.lvalue.place, kind);
          setKind(instr.lvalue, kind);
          break;
        }
        case 'ComputedLoad': {
          visitPlace(instr.value.object);
          const kind = getKindForPlace(instr.value.object);
          setKind(instr.lvalue, joinKinds(getKindForPlace(instr.lvalue), kind));
          break;
        }
        case 'PropertyLoad': {
          const objectKind = getKindForPlace(instr.value.object);
          const isHookProperty = isHookName(instr.value.property);
          let kind: Kind;
          switch (objectKind) {
            case Kind.Error: {
              kind = Kind.Error;
              break;
            }
            case Kind.KnownHook: {
              /**
               * const useFoo;
               * function Component() {
               *   let x = useFoo.useBar; // useFoo is KnownHook, any property from it inherits KnownHook
               * }
               */
              kind = isHookProperty ? Kind.KnownHook : Kind.Local;
              break;
            }
            case Kind.PotentialHook: {
              /**
               * function Component(props) {
               *   let useFoo;
               *   let x = useFoo.useBar; // useFoo is PotentialHook, any property from it inherits PotentialHook
               * }
               */
              kind = Kind.PotentialHook;
              break;
            }
            case Kind.Global: {
              /**
               * function Component() {
               *   let x = React.useState; // hook-named property of global is knownhook
               *   let y = React.foo; // else inherit Global
               * }
               */
              kind = isHookProperty ? Kind.KnownHook : Kind.Global;
              break;
            }
            case Kind.Local: {
              /**
               * function Component() {
               *   let o = createObject();
               *   let x = o.useState; // hook-named property of local is potentialhook
               *   let y = o.foo; // else inherit local
               * }
               */
              kind = isHookProperty ? Kind.PotentialHook : Kind.Local;
              break;
            }
            default: {
              assertExhaustive(objectKind, `Unexpected kind \`${objectKind}\``);
            }
          }
          setKind(instr.lvalue, kind);
          break;
        }
        case 'CallExpression': {
          const calleeKind = getKindForPlace(instr.value.callee);
          const isHookCallee =
            calleeKind === Kind.KnownHook || calleeKind === Kind.PotentialHook;
          if (isHookCallee && !unconditionalBlocks.has(block.id)) {
            recordConditionalHookError(instr.value.callee);
          } else if (calleeKind === Kind.PotentialHook) {
            recordDynamicHookUsageError(instr.value.callee);
          }
          /**
           * We intentionally skip the callee because it's validated above
           */
          for (const operand of eachInstructionOperand(instr)) {
            if (operand === instr.value.callee) {
              continue;
            }
            visitPlace(operand);
          }
          break;
        }
        case 'MethodCall': {
          const calleeKind = getKindForPlace(instr.value.property);
          const isHookCallee =
            calleeKind === Kind.KnownHook || calleeKind === Kind.PotentialHook;
          if (isHookCallee && !unconditionalBlocks.has(block.id)) {
            recordConditionalHookError(instr.value.property);
          } else if (calleeKind === Kind.PotentialHook) {
            recordDynamicHookUsageError(instr.value.property);
          }
          /*
           * We intentionally skip the property because it's validated above
           */
          for (const operand of eachInstructionOperand(instr)) {
            if (operand === instr.value.property) {
              continue;
            }
            visitPlace(operand);
          }
          break;
        }
        case 'Destructure': {
          visitPlace(instr.value.value);
          const objectKind = getKindForPlace(instr.value.value);
          for (const lvalue of eachInstructionLValue(instr)) {
            const isHookProperty =
              lvalue.identifier.name !== null &&
              isHookName(lvalue.identifier.name.value);
            let kind: Kind;
            switch (objectKind) {
              case Kind.Error: {
                kind = Kind.Error;
                break;
              }
              case Kind.KnownHook: {
                kind = Kind.KnownHook;
                break;
              }
              case Kind.PotentialHook: {
                kind = Kind.PotentialHook;
                break;
              }
              case Kind.Global: {
                kind = isHookProperty ? Kind.KnownHook : Kind.Global;
                break;
              }
              case Kind.Local: {
                kind = isHookProperty ? Kind.PotentialHook : Kind.Local;
                break;
              }
              default: {
                assertExhaustive(
                  objectKind,
                  `Unexpected kind \`${objectKind}\``,
                );
              }
            }
            setKind(lvalue, kind);
          }
          break;
        }
        case 'ObjectMethod':
        case 'FunctionExpression': {
          visitFunctionExpression(errors, instr.value.loweredFunc.func);
          break;
        }
        default: {
          /*
           * Else check usages of operands, but do *not* flow properties
           * from operands into the lvalues. For example, `let x = identity(y)`
           * does not infer `x` as a potential hook even if `y` is a potential hook.
           */
          for (const operand of eachInstructionOperand(instr)) {
            visitPlace(operand);
          }
          for (const lvalue of eachInstructionLValue(instr)) {
            const kind = getKindForPlace(lvalue);
            setKind(lvalue, kind);
          }
        }
      }
    }
    for (const operand of eachTerminalOperand(block.terminal)) {
      visitPlace(operand);
    }
  }

  for (const [, error] of errorsByPlace) {
    errors.push(error);
  }
  if (errors.hasErrors()) {
    throw errors;
  }
}

function visitFunctionExpression(errors: CompilerError, fn: HIRFunction): void {
  for (const [, block] of fn.body.blocks) {
    for (const instr of block.instructions) {
      switch (instr.value.kind) {
        case 'ObjectMethod':
        case 'FunctionExpression': {
          visitFunctionExpression(errors, instr.value.loweredFunc.func);
          break;
        }
        case 'MethodCall':
        case 'CallExpression': {
          const callee =
            instr.value.kind === 'CallExpression'
              ? instr.value.callee
              : instr.value.property;
          const hookKind = getHookKind(fn.env, callee.identifier);
          if (hookKind != null) {
            errors.pushErrorDetail(
              new CompilerErrorDetail({
                severity: ErrorSeverity.InvalidReact,
                reason:
                  'Hooks must be called at the top level in the body of a function component or custom hook, and may not be called within function expressions. See the Rules of Hooks (https://react.dev/warnings/invalid-hook-call-warning)',
                loc: callee.loc,
                description: `Cannot call ${hookKind} within a function component`,
                suggestions: null,
              }),
            );
          }
          break;
        }
      }
    }
  }
}

File: compiler/crates/react_compiler_validation/src/validate_hooks_usage.rs
// Copyright (c) Meta Platforms, Inc. and affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

//! Validates hooks usage rules.
//!
//! Port of ValidateHooksUsage.ts.
//! Ensures hooks are called unconditionally, not passed as values,
//! and not called dynamically. Also validates that hooks are not
//! called inside function expressions.

use std::collections::HashMap;

use indexmap::IndexMap;
use react_compiler_diagnostics::{
    CompilerDiagnostic, CompilerErrorDetail, ErrorCategory, SourceLocation,
};
use react_compiler_hir::{
    FunctionId, HirFunction, Identifier, IdentifierId,
    InstructionValue, ParamPattern, Place, PropertyLiteral,
    Type, visitors,
};
use react_compiler_hir::visitors::{each_pattern_operand, each_terminal_operand};
use react_compiler_hir::dominator::compute_unconditional_blocks;
use react_compiler_hir::environment::{is_hook_name, Environment};
use react_compiler_hir::object_shape::HookKind;

/// Value classification for hook validation.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum Kind {
    Error,
    KnownHook,
    PotentialHook,
    Global,
    Local,
}

fn join_kinds(a: Kind, b: Kind) -> Kind {
    if a == Kind::Error || b == Kind::Error {
        Kind::Error
    } else if a == Kind::KnownHook || b == Kind::KnownHook {
        Kind::KnownHook
    } else if a == Kind::PotentialHook || b == Kind::PotentialHook {
        Kind::PotentialHook
    } else if a == Kind::Global || b == Kind::Global {
        Kind::Global
    } else {
        Kind::Local
    }
}

fn get_kind_for_place(
    place: &Place,
    value_kinds: &HashMap<IdentifierId, Kind>,
    identifiers: &[Identifier],
) -> Kind {
    let known_kind = value_kinds.get(&place.identifier).copied();
    let ident = &identifiers[place.identifier.0 as usize];
    if let Some(ref name) = ident.name {
        if is_hook_name(name.value()) {
            return join_kinds(known_kind.unwrap_or(Kind::Local), Kind::PotentialHook);
        }
    }
    known_kind.unwrap_or(Kind::Local)
}

fn ident_is_hook_name(identifier_id: IdentifierId, identifiers: &[Identifier]) -> bool {
    let ident = &identifiers[identifier_id.0 as usize];
    if let Some(ref name) = ident.name {
        is_hook_name(name.value())
    } else {
        false
    }
}

fn get_hook_kind_for_id<'a>(
    identifier_id: IdentifierId,
    identifiers: &[Identifier],
    types: &[Type],
    env: &'a Environment,
) -> Result<Option<&'a HookKind>, CompilerDiagnostic> {
    let identifier = &identifiers[identifier_id.0 as usize];
    let ty = &types[identifier.type_.0 as usize];
    env.get_hook_kind_for_type(ty)
}

fn visit_place(
    place: &Place,
    value_kinds: &HashMap<IdentifierId, Kind>,
    errors_by_loc: &mut IndexMap<SourceLocation, CompilerErrorDetail>,
    env: &mut Environment,
) {
    let kind = value_kinds.get(&place.identifier).copied();
    if kind == Some(Kind::KnownHook) {
        record_invalid_hook_usage_error(place, errors_by_loc, env);
    }
}

fn record_conditional_hook_error(
    place: &Place,
    value_kinds: &mut HashMap<IdentifierId, Kind>,
    errors_by_loc: &mut IndexMap<SourceLocation, CompilerErrorDetail>,
    env: &mut Environment,
) {
    value_kinds.insert(place.identifier, Kind::Error);
    let reason = "Hooks must always be called in a consistent order, and may not be called conditionally. See the Rules of Hooks (https://react.dev/warnings/invalid-hook-call-warning)".to_string();
    if let Some(loc) = place.loc {
        let previous = errors_by_loc.get(&loc);
        if previous.is_none() || previous.unwrap().reason != reason {
            errors_by_loc.insert(
                loc,
                CompilerErrorDetail {
                    category: ErrorCategory::Hooks,
                    reason,
                    description: None,
                    loc: Some(loc),
                    suggestions: None,
                },
            );
        }
    } else {
        env.record_error(CompilerErrorDetail {
            category: ErrorCategory::Hooks,
            reason,
            description: None,
            loc: None,
            suggestions: None,
        });
    }
}

fn record_invalid_hook_usage_error(
    place: &Place,
    errors_by_loc: &mut IndexMap<SourceLocation, CompilerErrorDetail>,
    env: &mut Environment,
) {
    let reason = "Hooks may not be referenced as normal values, they must be called. See https://react.dev/reference/rules/react-calls-components-and-hooks#never-pass-around-hooks-as-regular-values".to_string();
    if let Some(loc) = place.loc {
        if !errors_by_loc.contains_key(&loc) {
            errors_by_loc.insert(
                loc,
                CompilerErrorDetail {
                    category: ErrorCategory::Hooks,
                    reason,
                    description: None,
                    loc: Some(loc),
                    suggestions: None,
                },
            );
        }
    } else {
        env.record_error(CompilerErrorDetail {
            category: ErrorCategory::Hooks,
            reason,
            description: None,
            loc: None,
            suggestions: None,
        });
    }
}

fn record_dynamic_hook_usage_error(
    place: &Place,
    errors_by_loc: &mut IndexMap<SourceLocation, CompilerErrorDetail>,
    env: &mut Environment,
) {
    let reason = "Hooks must be the same function on every render, but this value may change over time to a different function. See https://react.dev/reference/rules/react-calls-components-and-hooks#dont-dynamically-use-hooks".to_string();
    if let Some(loc) = place.loc {
        if !errors_by_loc.contains_key(&loc) {
            errors_by_loc.insert(
                loc,
                CompilerErrorDetail {
                    category: ErrorCategory::Hooks,
                    reason,
                    description: None,
                    loc: Some(loc),
                    suggestions: None,
                },
            );
        }
    } else {
        env.record_error(CompilerErrorDetail {
            category: ErrorCategory::Hooks,
            reason,
            description: None,
            loc: None,
            suggestions: None,
        });
    }
}

/// Validates hooks usage rules for a function.
pub fn validate_hooks_usage(func: &HirFunction, env: &mut Environment) -> Result<(), react_compiler_diagnostics::CompilerDiagnostic> {
    let unconditional_blocks = compute_unconditional_blocks(func, env.next_block_id().0)?;
    let mut errors_by_loc: IndexMap<SourceLocation, CompilerErrorDetail> = IndexMap::new();
    let mut value_kinds: HashMap<IdentifierId, Kind> = HashMap::new();

    // Process params
    for param in &func.params {
        let place = match param {
            ParamPattern::Place(p) => p,
            ParamPattern::Spread(s) => &s.place,
        };
        let kind = get_kind_for_place(place, &value_kinds, &env.identifiers);
        value_kinds.insert(place.identifier, kind);
    }

    // Process blocks
    for (_block_id, block) in &func.body.blocks {
        // Process phis
        for phi in &block.phis {
            let mut kind = if ident_is_hook_name(phi.place.identifier, &env.identifiers) {
                Kind::PotentialHook
            } else {
                Kind::Local
            };
            for (_, operand) in &phi.operands {
                if let Some(&operand_kind) = value_kinds.get(&operand.identifier) {
                    kind = join_kinds(kind, operand_kind);
                }
            }
            value_kinds.insert(phi.place.identifier, kind);
        }

        // Process instructions
        for &instr_id in &block.instructions {
            let instr = &func.instructions[instr_id.0 as usize];
            let lvalue_id = instr.lvalue.identifier;

            match &instr.value {
                InstructionValue::LoadGlobal { .. } => {
                    if get_hook_kind_for_id(lvalue_id, &env.identifiers, &env.types, env)?.is_some()
                    {
                        value_kinds.insert(lvalue_id, Kind::KnownHook);
                    } else {
                        value_kinds.insert(lvalue_id, Kind::Global);
                    }
                }
                InstructionValue::LoadContext { place, .. }
                | InstructionValue::LoadLocal { place, .. } => {
                    visit_place(place, &value_kinds, &mut errors_by_loc, env);
                    let kind = get_kind_for_place(place, &value_kinds, &env.identifiers);
                    value_kinds.insert(lvalue_id, kind);
                }
                InstructionValue::StoreLocal { lvalue, value, .. }
                | InstructionValue::StoreContext { lvalue, value, .. } => {
                    visit_place(value, &value_kinds, &mut errors_by_loc, env);
                    let kind = join_kinds(
                        get_kind_for_place(value, &value_kinds, &env.identifiers),
                        get_kind_for_place(&lvalue.place, &value_kinds, &env.identifiers),
                    );
                    value_kinds.insert(lvalue.place.identifier, kind);
                    value_kinds.insert(lvalue_id, kind);
                }
                InstructionValue::ComputedLoad { object, .. } => {
                    visit_place(object, &value_kinds, &mut errors_by_loc, env);
                    let kind = get_kind_for_place(object, &value_kinds, &env.identifiers);
                    let lvalue_kind =
                        get_kind_for_place(&instr.lvalue, &value_kinds, &env.identifiers);
                    value_kinds.insert(lvalue_id, join_kinds(lvalue_kind, kind));
                }
                InstructionValue::PropertyLoad {
                    object, property, ..
                } => {
                    let object_kind =
                        get_kind_for_place(object, &value_kinds, &env.identifiers);
                    let is_hook_property = match property {
                        PropertyLiteral::String(s) => is_hook_name(s),
                        PropertyLiteral::Number(_) => false,
                    };
                    let kind = match object_kind {
                        Kind::Error => Kind::Error,
                        Kind::KnownHook => {
                            if is_hook_property {
                                Kind::KnownHook
                            } else {
                                Kind::Local
                            }
                        }
                        Kind::PotentialHook => Kind::PotentialHook,
                        Kind::Global => {
                            if is_hook_property {
                                Kind::KnownHook
                            } else {
                                Kind::Global
                            }
                        }
                        Kind::Local => {
                            if is_hook_property {
                                Kind::PotentialHook
                            } else {
                                Kind::Local
                            }
                        }
                    };
                    value_kinds.insert(lvalue_id, kind);
                }
                InstructionValue::CallExpression { callee, args, .. } => {
                    let callee_kind =
                        get_kind_for_place(callee, &value_kinds, &env.identifiers);
                    let is_hook_callee =
                        callee_kind == Kind::KnownHook || callee_kind == Kind::PotentialHook;
                    if is_hook_callee && !unconditional_blocks.contains(&block.id) {
                        record_conditional_hook_error(
                            callee,
                            &mut value_kinds,
                            &mut errors_by_loc,
                            env,
                        );
                    } else if callee_kind == Kind::PotentialHook {
                        record_dynamic_hook_usage_error(callee, &mut errors_by_loc, env);
                    }
                    // Visit all operands except callee
                    for arg in args {
                        let place = match arg {
                            react_compiler_hir::PlaceOrSpread::Place(p) => p,
                            react_compiler_hir::PlaceOrSpread::Spread(s) => &s.place,
                        };
                        visit_place(place, &value_kinds, &mut errors_by_loc, env);
                    }
                }
                InstructionValue::MethodCall {
                    receiver,
                    property,
                    args,
                    ..
                } => {
                    let callee_kind =
                        get_kind_for_place(property, &value_kinds, &env.identifiers);
                    let is_hook_callee =
                        callee_kind == Kind::KnownHook || callee_kind == Kind::PotentialHook;
                    if is_hook_callee && !unconditional_blocks.contains(&block.id) {
                        record_conditional_hook_error(
                            property,
                            &mut value_kinds,
                            &mut errors_by_loc,
                            env,
                        );
                    } else if callee_kind == Kind::PotentialHook {
                        record_dynamic_hook_usage_error(
                            property,
                            &mut errors_by_loc,
                            env,
                        );
                    }
                    // Visit receiver and args (not property)
                    visit_place(receiver, &value_kinds, &mut errors_by_loc, env);
                    for arg in args {
                        let place = match arg {
                            react_compiler_hir::PlaceOrSpread::Place(p) => p,
                            react_compiler_hir::PlaceOrSpread::Spread(s) => &s.place,
                        };
                        visit_place(place, &value_kinds, &mut errors_by_loc, env);
                    }
                }
                InstructionValue::Destructure { lvalue, value, .. } => {
                    visit_place(value, &value_kinds, &mut errors_by_loc, env);
                    let object_kind =
                        get_kind_for_place(value, &value_kinds, &env.identifiers);
                    // Process instr.lvalue and all pattern operands (matching TS eachInstructionLValue)
                    let pattern_places = each_pattern_operand(&lvalue.pattern);
                    let all_lvalues = std::iter::once(instr.lvalue.clone())
                        .chain(pattern_places.into_iter());
                    for place in all_lvalues {
                        let is_hook_property =
                            ident_is_hook_name(place.identifier, &env.identifiers);
                        let kind = match object_kind {
                            Kind::Error => Kind::Error,
                            Kind::KnownHook => Kind::KnownHook,
                            Kind::PotentialHook => Kind::PotentialHook,
                            Kind::Global => {
                                if is_hook_property {
                                    Kind::KnownHook
                                } else {
                                    Kind::Global
                                }
                            }
                            Kind::Local => {
                                if is_hook_property {
                                    Kind::PotentialHook
                                } else {
                                    Kind::Local
                                }
                            }
                        };
                        value_kinds.insert(place.identifier, kind);
                    }
                }
                InstructionValue::ObjectMethod { lowered_func, .. }
                | InstructionValue::FunctionExpression { lowered_func, .. } => {
                    visit_function_expression(env, lowered_func.func);
                }
                _ => {
                    // For all other instructions: visit operands, set lvalue kinds
                    // Matches TS which uses eachInstructionOperand + eachInstructionLValue
                    visit_all_operands(
                        &instr.value,
                        &value_kinds,
                        &mut errors_by_loc,
                        env,
                    );
                    // Set kind for instr.lvalue
                    let kind =
                        get_kind_for_place(&instr.lvalue, &value_kinds, &env.identifiers);
                    value_kinds.insert(lvalue_id, kind);
                    // Also set kind for value-level lvalues (e.g. DeclareLocal, PrefixUpdate, PostfixUpdate)
                    for lv in visitors::each_instruction_value_lvalue(&instr.value) {
                        let lv_kind =
                            get_kind_for_place(&lv, &value_kinds, &env.identifiers);
                        value_kinds.insert(lv.identifier, lv_kind);
                    }
                }
            }
        }

        // Visit terminal operands
        for place in each_terminal_operand(&block.terminal) {
            visit_place(&place, &value_kinds, &mut errors_by_loc, env);
        }
    }

    // Record all accumulated errors (in insertion order, matching TS Map iteration)
    for (_, error_detail) in errors_by_loc {
        env.record_error(error_detail);
    }
    Ok(())
}

/// Visit a function expression to check for hook calls inside it.
/// Processes instructions in order, visiting nested functions immediately
/// (before processing subsequent calls) to match TS error ordering.
fn visit_function_expression(env: &mut Environment, func_id: FunctionId) {
    // Collect items in instruction order to process them sequentially.
    // Each item is either a call to check or a nested function to visit.
    enum Item {
        Call(IdentifierId, Option<SourceLocation>),
        NestedFunc(FunctionId),
    }

    let func = &env.functions[func_id.0 as usize];
    let mut items: Vec<Item> = Vec::new();

    for (_block_id, block) in &func.body.blocks {
        for &instr_id in &block.instructions {
            let instr = &func.instructions[instr_id.0 as usize];
            match &instr.value {
                InstructionValue::ObjectMethod { lowered_func, .. }
                | InstructionValue::FunctionExpression { lowered_func, .. } => {
                    items.push(Item::NestedFunc(lowered_func.func));
                }
                InstructionValue::CallExpression { callee, .. } => {
                    items.push(Item::Call(callee.identifier, callee.loc));
                }
                InstructionValue::MethodCall { property, .. } => {
                    items.push(Item::Call(property.identifier, property.loc));
                }
                _ => {}
            }
        }
    }

    // Process items in instruction order (matching TS which visits nested
    // functions immediately before processing subsequent calls)
    for item in items {
        match item {
            Item::Call(identifier_id, loc) => {
                let identifier = &env.identifiers[identifier_id.0 as usize];
                let ty = &env.types[identifier.type_.0 as usize];
                let hook_kind = env.get_hook_kind_for_type(ty).ok().flatten().cloned();
                if let Some(hook_kind) = hook_kind {
                    let description = format!(
                        "Cannot call {} within a function expression",
                        if hook_kind == HookKind::Custom {
                            "hook"
                        } else {
                            hook_kind_display(&hook_kind)
                        }
                    );
                    env.record_error(CompilerErrorDetail {
                        category: ErrorCategory::Hooks,
                        reason: "Hooks must be called at the top level in the body of a function component or custom hook, and may not be called within function expressions. See the Rules of Hooks (https://react.dev/warnings/invalid-hook-call-warning)".to_string(),
                        description: Some(description),
                        loc,
                        suggestions: None,
                    });
                }
            }
            Item::NestedFunc(nested_func_id) => {
                visit_function_expression(env, nested_func_id);
            }
        }
    }
}

fn hook_kind_display(kind: &HookKind) -> &'static str {
    match kind {
        HookKind::UseContext => "useContext",
        HookKind::UseState => "useState",
        HookKind::UseActionState => "useActionState",
        HookKind::UseReducer => "useReducer",
        HookKind::UseRef => "useRef",
        HookKind::UseEffect => "useEffect",
        HookKind::UseLayoutEffect => "useLayoutEffect",
        HookKind::UseInsertionEffect => "useInsertionEffect",
        HookKind::UseMemo => "useMemo",
        HookKind::UseCallback => "useCallback",
        HookKind::UseTransition => "useTransition",
        HookKind::UseImperativeHandle => "useImperativeHandle",
        HookKind::UseEffectEvent => "useEffectEvent",
        HookKind::UseOptimistic => "useOptimistic",
        HookKind::Custom => "hook",
    }
}

/// Visit all operands of an instruction value (generic fallback).
/// Uses the canonical `each_instruction_value_operand` from visitors.
fn visit_all_operands(
    value: &InstructionValue,
    value_kinds: &HashMap<IdentifierId, Kind>,
    errors_by_loc: &mut IndexMap<SourceLocation, CompilerErrorDetail>,
    env: &mut Environment,
) {
    let operands = visitors::each_instruction_value_operand(value, &*env);
    for place in &operands {
        visit_place(place, value_kinds, errors_by_loc, env);
    }
}

